File: client/src/features/reader/readingPreparation.ts
// §22 « Préparer la lecture » from the reader: the page on screen goes to the home computer, which punctuates it for the
// voice; the prepared text comes back with the synchronization (asked every 10 s for 3 minutes).
import { useCallback, useEffect, useRef, useState } from 'react';
import { prepareReading } from '../../api/documents';
import { ApiError } from '../../api/http';
import { tts } from '../../i18n/fr/tts';
import { syncNow } from '../../sync/SyncEngine';
import type { PageModel } from './model';

const t = tts.prepare;

export type PreparationStatus = 'idle' | 'sending' | 'waiting' | 'ready';

export const PREPARATION_POLL_MS = 10_000;
export const PREPARATION_WAIT_MS = 3 * 60_000;

/** The voice reads at least one sentence of the page from a preparation. */
export function isPagePrepared(model: PageModel | undefined): boolean {
  return model?.blocks.some((b) => b.sentences.some((s) => s.spoken !== null)) ?? false;
}

export interface PreparationNotify {
  info(message: string): void;
  success(message: string): void;
  warning(message: string): void;
}

export function useReadingPreparation(
  documentId: string, pageIndex: number | null, model: PageModel | undefined, notify: PreparationNotify,
): { status: PreparationStatus; prepare(): void } {
  const [sending, setSending] = useState(false);
  const [waiting, setWaiting] = useState<{ pageIndex: number; since: number } | null>(null);
  const notifyRef = useRef(notify);
  notifyRef.current = notify;
  const prepared = isPagePrepared(model);

  // The prepared text arrived for the page we wait for.
  useEffect(() => {
    if (waiting !== null && waiting.pageIndex === pageIndex && prepared) {
      setWaiting(null);
      notifyRef.current.success(t.ready);
    }
  }, [waiting, pageIndex, prepared]);

  // Ask for the result with the synchronization, for a while.
  useEffect(() => {
    if (waiting === null) return undefined;
    const poll = setInterval(() => void syncNow(), PREPARATION_POLL_MS);
    const stop = setTimeout(() => {
      setWaiting(null);
      notifyRef.current.info(t.later);
    }, Math.max(0, waiting.since + PREPARATION_WAIT_MS - Date.now()));
    return () => {
      clearInterval(poll);
      clearTimeout(stop);
    };
  }, [waiting]);

  const prepare = useCallback((): void => {
    if (pageIndex === null || sending) return;
    const page = pageIndex;
    setSending(true);
    void prepareReading(documentId, { pageIndexes: [page] })
      .then((response) => {
        if (response.unavailable !== null) {
          notifyRef.current.info(t.unavailable);
          return;
        }
        notifyRef.current.info(response.queued > 0 ? t.queued : t.already);
        setWaiting({ pageIndex: page, since: Date.now() });
      })
      .catch((error: unknown) => {
        notifyRef.current.warning(error instanceof ApiError && error.code === 'offline' ? t.offline : t.failed);
      })
      .finally(() => setSending(false));
  }, [documentId, pageIndex, sending]);

  const status: PreparationStatus = sending
    ? 'sending'
    : prepared
      ? 'ready'
      : waiting !== null && waiting.pageIndex === pageIndex
        ? 'waiting'
        : 'idle';
  return { status, prepare };
}

File: ios-native/CarnetMalin/Screens/Reader/ReadingPreparation.swift
import CarnetKit
import Foundation

/// §22 « Préparer la lecture » from the reader: the page on screen goes to the home computer, which punctuates it for
/// the voice. The prepared text comes back with the synchronization, asked every 10 s for 3 minutes.
///
/// Only the page on screen is asked for: a child may ask for one or two pages, the whole book is the parent's call.
@MainActor
final class ReadingPreparation: ObservableObject {
    enum Status: Equatable {
        case idle, sending, waiting, ready
    }

    static let pollInterval: Duration = .seconds(10)
    static let waitLimit: TimeInterval = 3 * 60

    @Published private(set) var sending = false
    /// The page we wait for, and since when.
    @Published private(set) var waiting: (pageIndex: Int, since: Date)?

    /// Holds the model weakly: a closed book stops asking at the next tick.
    private var poll: Task<Void, Never>?

    func status(pageIndex: Int, prepared: Bool) -> Status {
        if sending { return .sending }
        if prepared { return .ready }
        if let waiting, waiting.pageIndex == pageIndex { return .waiting }
        return .idle
    }

    /// Sends the page. `sync` runs a synchronization and gives the reader its new pages; `notify` shows a message.
    func prepare(
        api: APIClient, documentId: String, pageIndex: Int,
        sync: @escaping @MainActor () async -> Void,
        isPrepared: @escaping @MainActor () -> Bool,
        notify: @escaping @MainActor (String, AppModel.Banner.Tone) -> Void
    ) {
        guard !sending else { return }
        sending = true
        Task {
            defer { sending = false }
            do {
                let answer = try await api.prepareReading(documentId: documentId, pageIndexes: [pageIndex])
                if answer.unavailable != nil {
                    notify(FR.TTS.Prepare.unavailable, .info)
                    return
                }
                notify(answer.queued > 0 ? FR.TTS.Prepare.queued : FR.TTS.Prepare.already, .info)
                wait(for: pageIndex, sync: sync, isPrepared: isPrepared, notify: notify)
            } catch APIError.offline {
                notify(FR.TTS.Prepare.offline, .warning)
            } catch {
                notify(FR.TTS.Prepare.failed, .warning)
            }
        }
    }

    private func wait(
        for pageIndex: Int,
        sync: @escaping @MainActor () async -> Void,
        isPrepared: @escaping @MainActor () -> Bool,
        notify: @escaping @MainActor (String, AppModel.Banner.Tone) -> Void
    ) {
        poll?.cancel()
        let since = Date()
        waiting = (pageIndex, since)
        poll = Task { [weak self] in
            while !Task.isCancelled {
                try? await Task.sleep(for: Self.pollInterval)
                guard !Task.isCancelled, let self else { return }
                await sync()
                if isPrepared() {
                    self.waiting = nil
                    notify(FR.TTS.Prepare.ready, .success)
                    return
                }
                if Date().timeIntervalSince(since) >= Self.waitLimit {
                    self.waiting = nil
                    notify(FR.TTS.Prepare.later, .info)
                    return
                }
            }
        }
    }

    /// The book is closed: nothing more to wait for here. The page is still prepared on the server and arrives with
    /// the next sync.
    func stop() {
        poll?.cancel()
        poll = nil
        waiting = nil
    }
}
